src/fsm.cpp

```diff
@@ -171,3 +179,114 @@ void fsm::FSM::validate_final_states() const {
     }
 }
 
+bool fsm::FSM::evaluate(const char* input) {
+    fsm::String word(input);
+
+    for(int i = 0, l = word.size(); i < l; i++){
+        FSM::transition(word[i] - '0'); // convert char to int
+    }
+
+    bool flag = FSM::is_in_final_state();
+    FSM::restart();
+
+    return flag;
+}
+
+fsm::FSM fsm::FSM::operator!() const {
+    std::vector<fsm::State> newFinalStates;
+    fsm::FSM complementMachine = *this;
+
+    for(int i = 0, sz = get_states_count(); i < sz; i++){
+        if(std::find(final_states_.begin(),final_states_.end(),states_[i]) == final_states_.end()){
+            newFinalStates.push_back(states_[i]);
+        }
+    }
+
+    complementMachine.set_final_states(newFinalStates);
+
+    return complementMachine;
+}
+
+fsm::FSM fsm::FSM::operator|(const fsm::FSM &rhs) const {
+    fsm::FSM unionMachine, thisMachine = *this, otherMachine = rhs;
+    thisMachine.restart();
+    otherMachine.restart();
+
+    fsm::State comboState = thisMachine.get_initial_state() + otherMachine.get_initial_state();  // Get the initial state.
+
+    auto alphabet = get_alphabet();
+    for(int i = 0, sz = get_alphabet_count(); i < sz; i++) { unionMachine.add_symbol(alphabet[i]); }  // Add the alphabet.
+    unionMachine.add_state(comboState);
+
+    fsm::fill(thisMachine, otherMachine, unionMachine, comboState);
+
+    unionMachine.set_initial_state(comboState);
+    unionMachine.restart();
+
+    return unionMachine;
+}
+
+void fsm::fill(fsm::FSM m1, fsm::FSM m2, fsm::FSM &m3, fsm::State prevState) {
+    for(int i = 0, sz = m1.get_alphabet_count(); i < sz; i++){
+        auto m1old = m1, m2old = m2;
+        auto m3States = m3.get_states();
+        char currLetter = m3.get_alphabet()[i];
+
+        m1.transition(currLetter);
+        m2.transition(currLetter);
+
+        fsm::State comboState = m1.get_current_state() + m2.get_current_state();
+
+        if(std::find(m3States.begin(), m3States.end(), comboState) == m3States.end()){
+            m3.add_state(comboState);
+            if(m1.is_in_final_state() || m2.is_in_final_state()) { m3.add_final_state(comboState); }
+            fsm::fill(m1, m2, m3, comboState);
+        }
+        m3.add_transition_rule(prevState, currLetter, comboState);
+        m1 = m1old;
+        m2 = m2old;
+    }
+}
+
+
+fsm::FSM fsm::FSM::operator&(const fsm::FSM &rhs) const {
+    std::vector<fsm::State> newEndStates;
+    fsm::FSM intersectionMachine = *this | rhs;
+    auto thisFinalStates = get_final_states();
+    auto otherFinalStates = rhs.get_final_states();
+    auto unionEndStates = intersectionMachine.get_final_states();
+
+    for(int i = 0, sz1 = get_final_states_count(); i < sz1; i++){
+        for(int j = 0, sz2 = rhs.get_final_states_count(); j < sz2; j++){
+            fsm::State doubleEndState = thisFinalStates[i] + otherFinalStates[j];
+            if(std::find(unionEndStates.begin(), unionEndStates.end(), doubleEndState) != unionEndStates.end()){
```

Comment:
Every end state of the union machine is made out of two "sub-states", of which at least one is in an end state in the operands. The loop only gets the end states, which "sub-states" are both end states.

ex. if A | B = C and q1, q4 are end states in A, B respectively, then all of these [q1q6, q1q4, q2q4] are endstates in C.
To get the intersection machine we just need get the states from the union machine that are made out of two "end sub-states", aka [q1q4].

The loop does exactly this, it constructs all possible endstates that are made of 2 "end sub-states", and checks if they are already pressent. If they exist, note them down. In a way the IF statement might not be needed, but it could be possible that an end state can be constructed, but is a totaly new state => we do not have it covered in the transition table.

I updated the code to be a bit more clear.